src/coreclr/tools/Common/TypeSystem/Common/MethodDesc.Async.cs

```diff
@@ -0,0 +1,183 @@
+        //
+        // When we see a Task-returning method we create 2 method variants that logically match the same method definition.
+        // One variant has the same signature/callconv as the defining method and another is a matching Async variant.
+        // Depending on whether the definition was a runtime async method or an ordinary Task-returning method,
+        // the IL body belongs to one of the variants and another variant is a synthetic thunk.
+        //
+        // The signature of the Async variant is derived from the original signature by replacing Task return type with
+        // modreq'd element type:
+        //   Example: "Task<int> Foo();"  ===> "modreq(Task`) int Foo();"
+        //   Example: "ValueTask Bar();"  ===> "modreq(ValueTask) void Bar();"
+        //
+        // The reason for this encoding is that:
+        //   - it uses parts of original signature, as-is, thus does not need to look for or construct anything
+        //   - it "unwraps" the element type.
+        //   - it is reversible. In particular nonconflicting signatures will map to nonconflicting ones.
+        //
+        // Async methods are called with CORINFO_CALLCONV_ASYNCCALL call convention.
+        //
+        // It is possible to get from one variant to another via GetAsyncOtherVariant.
+        //
+        // NOTE: not all Async methods are "variants" from a pair, see AsyncExplicitImpl below.
+        //=============================================================
+
+        // The following methods use special calling convention (CORINFO_CALLCONV_ASYNCCALL)
+        // These methods are emitted by the JIT as resumable state machines and also take an extra
+        // parameter and extra return - the continuation object.
+
+        // Async methods with actual IL implementation of a MethodImpl::Async method.
+        AsyncVariantImpl,
+
+        // Async methods with synthetic bodies that forward to a TaskReturning method.
+        AsyncVariantThunk,
+
+        // Methods that are explicitly declared as Async in metadata while not Task returning.
+        // This is a special case used in a few infrastructure methods like `Await`.
+        // Such methods do not get non-Async variants/thunks and can only be called from another Async method.
+        // NOTE: These methods have the original signature and it is not possible to tell if the method is Async
+        //       from the signature alone, thus all these methods are also JIT intrinsics.
+        AsyncExplicitImpl,
+    }
+
+    public abstract partial class MethodDesc : TypeSystemEntity
+    {
+        public virtual AsyncMethodKind AsyncMethodKind
+        {
+            get
+            {
+                return AsyncMethodKind.NotAsync;
+            }
+        }
+
+        public bool IsTaskReturning
+        {
+            get
+            {
+                return Signature.ReturnsTaskOrValueTask;
+            }
+        }
```

Comment:
We don't introduce these kinds of macros. We don't even have `MethodDesc.IsStatic`. `foo.Signature.ReturnsTaskOrValueTask` is fine.